### PR TITLE
Move GlobalCheckpointTracker and remove SequenceNumbersService

### PR DESCRIPTION
**Pull request by ywelsch**
This PR moves GlobalCheckpointTracker from the engine to IndexShard, where it better fits logically: Tracking the global checkpoint based on the local checkpoints of all shards in the replication group is not a property of the engine, but rather a property fulfilled by the current primary shard. The LocalCheckpointTracker on the other hand is driven by the contents of the local translog. By moving GlobalCheckpointTracker to IndexShard, it makes little sense to keep the SequenceNumbersService class around - it would only wrap the LocalCheckpointTracker. This PR therefore removes the class and replaces occurrences of SequenceNumbersService in the engine directly by LocalCheckpointTracker.

### CODE REVIEW

**Comment by bleskes** [comment]
I like this very much. I think we should do it but also take it a step further and total isolate the GlobalCheckpointTracker from the engine. Something like this, wdyt?

[patch.txt](https://github.com/elastic/elasticsearch/files/1562689/patch.txt)

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -237,6 +235,22 @@ public InternalEngine(EngineConfig engineConfig) {
         logger.trace("created new InternalEngine");
     }
 
+    private LocalCheckpointTracker createLocalCheckpointTracker(
+        BiFunction<Long, Long, LocalCheckpointTracker> localCheckpointTrackerSupplier) throws IOException {
+        final long maxSeqNo;
+        final long localCheckpoint;
+        if (openMode == EngineConfig.OpenMode.CREATE_INDEX_AND_TRANSLOG) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: paranoia - can you use a switch statement with a fallback logic and an exception on default?

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 2]
```diff
@@ -1321,7 +1321,16 @@ private void internalPerformTranslogRecovery(boolean skipTranslogRecovery, boole
 
         assert indexExists == false || assertMaxUnsafeAutoIdInCommit();
 
+
         final EngineConfig config = newEngineConfig(openMode);
+
+        if (openMode == EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: can you add a comment why it's important to do it before we  open the engine?

File: core/src/main/java/org/elasticsearch/index/store/Store.java [hunk 3]
```diff
@@ -211,17 +212,14 @@ private static SegmentInfos readSegmentsInfo(IndexCommit commit, Directory direc
...
-     * externally as it is not stored in the commit point.
+     * Loads the maximum sequence number and local checkpoint from the latest Lucene commit point.
      *
-     * @param globalCheckpoint the provided global checkpoint
-     * @return an instance of {@link SeqNoStats} populated with the local and global checkpoints, and the maximum sequence number
+     * @return a tuple populated with the maximum sequence number and the local checkpoint
      * @throws IOException if an I/O exception occurred reading the latest Lucene commit point from disk
      */
-    public SeqNoStats loadSeqNoStats(final long globalCheckpoint) throws IOException {
+    public Tuple<Long, Long> loadSeqNoStats() throws IOException {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/store/Store.java]: maybe call this load seqNoInfo? This way people won't be confused about SeqNoStats?

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 4]
```diff
@@ -461,7 +462,7 @@ TranslogWriter createWriter(long fileGeneration) throws IOException {
      *                              needed to solve and initialization problem while constructing an empty translog.
      *                              With no readers and no current, a call to  {@link #getMinFileGeneration()} would not work.
      */
-    private TranslogWriter createWriter(long fileGeneration, long initialMinTranslogGen) throws IOException {
+    private TranslogWriter createWriter(long fileGeneration, long initialMinTranslogGen, long initialGlobalCheckpoint) throws IOException {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: java doc update?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 5]
```diff
@@ -132,10 +132,21 @@ public static TranslogWriter create(ShardId shardId, String translogUUID, long f
             writeHeader(out, ref);
             channel.force(true);
             final Checkpoint checkpoint = Checkpoint.emptyTranslogCheckpoint(firstOperationOffset, fileGeneration,
-                globalCheckpointSupplier.getAsLong(), initialMinTranslogGen);
+                initialGlobalCheckpoint, initialMinTranslogGen);
             writeCheckpoint(channelFactory, file.getParent(), checkpoint);
+            final LongSupplier writerGlobalCheckpointSupplier;
+            if (Assertions.ENABLED) {
+                writerGlobalCheckpointSupplier = () -> {
+                    long gcp = globalCheckpointSupplier.getAsLong();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: 👍

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 6]
```diff
@@ -1015,40 +1001,32 @@ public void testCommitAdvancesMinTranslogForRecovery() throws IOException {
...
-                seqNoStats.getGlobalCheckpoint()) {
-                @Override
-                public long getGlobalCheckpoint() {
-                    return inSync.get() ? getLocalCheckpoint() : SequenceNumbers.UNASSIGNED_SEQ_NO;
-                }
-            };
-        engine = new InternalEngine(config(defaultSettings, store, translogPath, newMergePolicy(), null, null), seqNoServiceSupplier);
+        final AtomicLong globalCheckpoint = new AtomicLong(SequenceNumbers.UNASSIGNED_SEQ_NO);
+        final LongSupplier globalCheckpointSupplier = () -> globalCheckpoint.get();
+        engine = new InternalEngine(config(defaultSettings, store, translogPath, newMergePolicy(), null, null, globalCheckpointSupplier));
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: nice

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 7]
```diff
@@ -2035,20 +2002,14 @@ public void testSeqNoAndCheckpoints() throws IOException {
...
-            assertThat(initialEngine.seqNoService().stats().getLocalCheckpoint(), equalTo(primarySeqNo));
-            assertThat(initialEngine.seqNoService().stats().getGlobalCheckpoint(), equalTo(replicaLocalCheckpoint));
+            assertEquals(primarySeqNo, initialEngine.getLocalCheckpointTracker().getMaxSeqNo());
+            assertEquals(primarySeqNo, initialEngine.getLocalCheckpointTracker().getCheckpoint());
 
             assertThat(
                 Long.parseLong(initialEngine.commitStats().getUserData().get(SequenceNumbers.LOCAL_CHECKPOINT_KEY)),
                 equalTo(localCheckpoint));
             initialEngine.getTranslog().sync(); // to guarantee the global checkpoint is written to the translog checkpoint
-            assertThat(
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: why is this gone?

File: test/framework/src/main/java/org/elasticsearch/index/engine/EngineTestCase.java [hunk 8]
```diff
@@ -164,7 +164,9 @@ public EngineConfig copy(EngineConfig config, EngineConfig.OpenMode openMode, An
                 new CodecService(null, logger), config.getEventListener(), config.getQueryCache(), config.getQueryCachingPolicy(),
                 config.getForceNewHistoryUUID(), config.getTranslogConfig(), config.getFlushMergesAfter(),
                 config.getExternalRefreshListener(), Collections.emptyList(), config.getIndexSort(), config.getTranslogRecoveryRunner(),
-                config.getCircuitBreakerService());
+                config.getCircuitBreakerService(), config.getGlobalCheckpointSupplier());
+                //new GlobalCheckpointTracker(config.getShardId(), config.getAllocationId(),
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/index/engine/EngineTestCase.java]: left overs?

File: test/framework/src/main/java/org/elasticsearch/index/shard/IndexShardTestCase.java [hunk 9]
```diff
@@ -633,4 +634,8 @@ protected void snapshotShard(final IndexShard shard,
     public static Engine getEngine(IndexShard indexShard) {
         return indexShard.getEngine();
     }
+
+    public static GlobalCheckpointTracker getGlobalCheckpointTracker(IndexShard indexShard) {
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/index/shard/IndexShardTestCase.java]: 👍

**Review by bleskes — APPROVED** [review]
LGTM. Awesome stats.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 1 above)
Comment (ywelsch) [reply core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: done in ea23a0d

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 2 above)
Comment (ywelsch) [reply core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: sure, added in ea23a0d

File: core/src/main/java/org/elasticsearch/index/store/Store.java (hunk 3 above)
Comment (ywelsch) [reply core/src/main/java/org/elasticsearch/index/store/Store.java]: ok

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java (hunk 4 above)
Comment (ywelsch) [reply core/src/main/java/org/elasticsearch/index/translog/Translog.java]: ok

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java (hunk 7 above)
Comment (ywelsch) [reply core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: I found it tricky to keep these checks in the test with the other changes around Engine not exposing GlobalCheckpointTracker. I also did not like that this test was checking a mixture between local checkpoint and global checkpoint properties (where the global checkpoint properties are mostly trivial, and covered by GlobalCheckpointTrackerTests). I've found a way to reintroduce these checks though (see ea23a0d)

File: test/framework/src/main/java/org/elasticsearch/index/engine/EngineTestCase.java (hunk 8 above)
Comment (ywelsch) [reply test/framework/src/main/java/org/elasticsearch/index/engine/EngineTestCase.java]: right -> removed

**Comment by ywelsch** [comment]
backport pending waiting on backport of https://github.com/elastic/elasticsearch/pull/27826

**Comment by dnhatn** [comment]
Backported 6.x in https://github.com/elastic/elasticsearch/commit/1576a2b3a4ebcdd2fb49ede1cbcdd99fabbb890e

**Pull request merged**